Implement subgraph unpacking

AustinMroz opened the pull request:
![unpacking](https://github.com/user-attachments/assets/914c66a8-2092-4a57-bd5b-bea146ad2878)

Known Issues:
- [x] ~~Reroutes on links connecting to the subgraph node are left dangling~~
- [x] ~~Widget related errors when destorying a subgraph node nested inside another subgraph.~~

Resolves #4866 by changing the convert to subgraph button to an unpack button

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-4840-Implement-subgraph-unpacking-2496d73d365081668119e83072e8211f) by [Unito](https://www.unito.io)

AustinMroz committed:
Implement subgraph unpacking

AustinMroz committed:
After unpacking subgraph, select unpacked nodes

JorgeR81 commented:
> Re-arrange nodes in parent graph to make room for unpacked nodes.

This would be nice, but it may be hard to make it work correctly every time. 
We could have too many nodes close together, in a large workflow.
Or they could be inside groups, and these would need to be re-scaled too.
Or we may just not like the new automatic node placement. 

Could it be possible to show some sort of preview ( e.g. translucid blue ) of the unpacked nodes ?
This would allow the user to make room manually, before unpacking.

AustinMroz commented:
I agree on both points. My original thought for "making room" was to always push existing parent graph nodes left or  right and ignore height. It would be comparatively simple to implement, but rather disruptive. I had also failed to consider how to handle groups (the cosmetic colored background rectangles). I now lean towards not including functionality for making room.

Selecting unpacked nodes has been implemented and at least gives the option of dragging the new nodes to a new corner of the graph for manually arrange later. Similarly, you can Ctrl+z undo the unpack operation if it winds up being too much of a mess.

I had considered and like the idea of previewing where nodes will unpack to. I don't currently have a solid grasp of how to implement it in a performant or intuitive manner. Maybe as a future PR.

Thanks for the feedback!

JorgeR81 commented:
> I had considered and like the idea of previewing where nodes will unpack to. I don't currently have a solid grasp of how to implement it in a performant or intuitive manner. Maybe as a future PR.

Some thoughts for when you have time to work on this.

- Maybe it's not necessary to show the actual nodes, just a single rectangular outline, of the space they would take. 

- The outline could show up when we mouseover the right-click option to unpack. 

AustinMroz committed:
Revoke previews before unpacking subgraph

AustinMroz committed:
Add simple tests, initial reroute support

Tests are basic and will be expanded later

Pre-existing bugs with subgraphs and reroutes make verifying
implementation difficult, so this functionality will be returned to
later.

AustinMroz committed:
Fix reroute migration

Applying position offsets results in reroutes being incorrectly offset,
so this step is currently skipped. Further investigation is required.

AustinMroz committed:
Fix offsetting

Offsetting on nodes was being reset at the configure step, causing the
earlier noted discrepancy of positioning.

Performing any positioning at all is opinionated and deserving of
further review. It causes the unpack operation to not be a true inverse
of converting to subgraph.

I consider minor repositioning for short-lived subgraphs to be a lesser
evil compared to nodes disappearing into a far corner of the canvas for
long lived subgraphs.

AustinMroz committed:
Allow unpacking subgraphs from selection toolbox

When the selection consists of a single subgraph node, the convert to
subgraph button in the selection toolbox changes to an unpack subgraph
button.

Known Issue: Clicking the convert to subgraph button does not cause the
button to swap to unpacking immediately. This is a pre-existing bug
since other buttons like "open mask editor" also fail to update and
rather outside my area of expertise.

AustinMroz committed:
Add further unpacking tests

Fixes a prior test returning early

Adds tests for reroute capabilities. The second of these tests has
exposed a bug with reroutes on links connected to the unpacked subgraph
node being left floating

AustinMroz committed:
Fix incorrect preview revocation code

Also swaps to the dedicated isRootGraph function

christian-byrne commented:
Small conflict because I also added a command in a separate PR.

AustinMroz committed:
Fix unpacking subgraphs nested inside subgraphs

AustinMroz commented:
Looks like github really didn't like the merge direction.  I can reverse the direction and force-push if need be.

AustinMroz committed:
Rewrite reroute handling

The prior method for handling reroutes failed to properly handle
reroutes external to the unwrapped subgraph. The implmentation has been
changed to one that instead seperates the process of creating and
linking reroutes.

AustinMroz committed:
Remove stale links from node info when unpacking

Failure to do so would resulted in infrequent, unintended disconnection
when unpacking.

AustinMroz committed:
Fix input reroute test

Remove logging

christian-byrne reviewed (approved):
LGTM, very nice!
